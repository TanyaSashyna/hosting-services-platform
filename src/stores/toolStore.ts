import { observable } from "mobx";

import { ToolApi, StatusTool, DnsChangesType } from "../api/ToolApi";

export class ToolStore {
  @observable domain = "";
  @observable status: StatusTool = "disabled";
  @observable dnsChangesRequired = false;
  @observable dnsChanges: DnsChangesType[] = [];
  @observable hostReachable = false;
  @observable canChangeDns = true;
  @observable lastCacheClearedAt = "";

  async getToolStatus(id: string): Promise<void> {
    this.domain = "";
    this.status = "disabled";
    const data = await ToolApi.getToolStatus(id);
    this.status = data.status;
    this.canChangeDns = data.canChangeDns;
    this.dnsChangesRequired = data.dnsChangesRequired;
    this.dnsChanges = data.dnsChanges;
    this.hostReachable = data.hostReachable;
    this.domain = data.domain;
    this.lastCacheClearedAt = data.lastCacheClearedAt || "";
  }

  async addTool(id: string): Promise<void> {
    this.status = "inProgress";
    await ToolApi.addTool(id);
  }

  async cancelTool(id: string): Promise<void> {
    await ToolApi.cancelTool(id);
    this.status = "disabled";
  }

  async clearCache(id: string): Promise<void> {
    const data = await ToolApi.clearCache(id);
    this.lastCacheClearedAt = data.lastCacheClearedAt;
  }
}

const store = new ToolStore();
export default store;
