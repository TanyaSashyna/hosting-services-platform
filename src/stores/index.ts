import toolStore from "./toolStore";

const stores = { toolStore };

export default stores;
