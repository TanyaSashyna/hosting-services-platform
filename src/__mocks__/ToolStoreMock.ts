/* eslint-disable @typescript-eslint/no-explicit-any  */
/* eslint-disable @typescript-eslint/no-empty-function  */

import { ToolStore } from "../stores/ToolStore";
import { StatusTool, DnsChangesType } from "../api/ToolApi";

class ToolStoreMock extends ToolStore {
  getToolStatus(): any {}
  addTool(): any {}
  cancelTool(): any {}
  status: StatusTool = "disabled";
  domain = "";
  dnsChangesRequired = true;
  dnsChanges: DnsChangesType[] = [];
  percentage = 0;
  hostReachable = true;
}

const toolStoreMock = new ToolStoreMock();

toolStoreMock.getToolStatus = jest.fn();
toolStoreMock.addTool = jest.fn();
toolStoreMock.cancelTool = jest.fn();

export default toolStoreMock;
