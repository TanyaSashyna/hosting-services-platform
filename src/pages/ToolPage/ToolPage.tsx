import { inject, observer } from "mobx-react";
import React, { FC, useEffect } from "react";
import {
  Page,
  useWindowTitle,
  useLoadingPage,
  useNavigate,
  useParams,
} from "@project-name-frontend/shared-context";

import { ToolStore } from "../../stores/toolStore";
import ToolContainer from "../../containers/ToolContainer";

type InjectedProps = {
  toolStore: ToolStore;
};

export const ToolPage: FC<InjectedProps> = ({ toolStore }) => {
  useWindowTitle("Tool");
  const navigate = useNavigate();
  const { id } = useParams();
  const { pushLoading, popLoading } = useLoadingPage();

  useEffect(() => {
    if (id) {
      pushLoading();
      toolStore.getToolStatus(id).finally(popLoading);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id]);

  const { domain } = toolStore;

  return (
    <Page
      layoutType="veryNarrow"
      title="Tool"
      subTitle={domain}
      onBack={(): void =>
        navigate(`/tool/overview/${id}`)
      }
      backText="Back"
    >
      {domain && (
        <ToolContainer id={id as string} />
      )}
    </Page>
  );
};

ToolPage.displayName = "ToolPage";

export default inject("toolStore")(observer(ToolPage)) as unknown as FC;
