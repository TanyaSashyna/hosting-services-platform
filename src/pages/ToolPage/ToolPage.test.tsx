import React from "react";
import { cleanup, render, fireEvent, waitFor } from "@testing-library/react";
import { Page } from "@project-name-frontend/ui-package";
import { Provider } from "mobx-react";
import { ToolPage } from "./ToolPage";
import ToolStoreMock from "../../__mocks__/ToolStoreMock";
import stores from "../../stores";

const mockedUsedNavigate = jest.fn();

jest.mock("@project-name-frontend/shared-context", () => ({
  ...jest.requireActual("@project-name-frontend/shared-context"),
  useNavigate: (): unknown => mockedUsedNavigate,
}));

afterEach(cleanup);

describe("ToolPage", () => {
  it("renders and matches snapshot", () => {
    ToolStoreMock.getToolStatus = jest.fn().mockResolvedValue({});

    const { container, asFragment } = render(
      <ToolPage toolStore={ToolStoreMock} />
    );
    expect(container).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it("triggers getToolStatus", () => {
    ToolStoreMock.getToolStatus = jest.fn();

    const { container } = render(<ToolPage toolStore={ToolStoreMock} />);

    waitFor(() => expect(ToolStoreMock.getToolStatus).toHaveBeenCalled());
    expect(container).toBeDefined();
  });

  it("triggers onBack", () => {
    ToolStoreMock.getToolStatus = jest.fn().mockResolvedValue({});
    ToolStoreMock.domain = "example.com";

    const { container } = render(
      <Provider {...stores}>
        <Page
          baseHeader="Tool"
          onBack={(): void => mockedUsedNavigate("/tool/overview/1")}
          backText={<button className="header__back">Back</button>}
        >
          <ToolPage toolStore={ToolStoreMock} />
        </Page>
      </Provider>
    );

    const backBtn = container.querySelector(".header__back") as Element;
    fireEvent.click(backBtn);

    expect(mockedUsedNavigate).toHaveBeenCalled();
  });
});
