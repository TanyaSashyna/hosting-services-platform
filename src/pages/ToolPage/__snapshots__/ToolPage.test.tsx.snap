// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolPage renders and matches snapshot 1`] = `<DocumentFragment />`;
